ops/model.py

```diff
@@ -2998,6 +3055,44 @@ def secret_remove(self, id: str, *, revision: Optional[int] = None):
             args.extend(['--revision', str(revision)])
         self._run_for_secret('secret-remove', *args)
 
+    def open_port(self, protocol: str, port: Optional[int] = None):
+        arg = f'{port}/{protocol}' if port is not None else protocol
+        self._run('open-port', arg)
+
+    def close_port(self, protocol: str, port: Optional[int] = None):
+        arg = f'{port}/{protocol}' if port is not None else protocol
+        self._run('close-port', arg)
+
+    def opened_ports(self) -> Set[OpenedPort]:
+        # We could use "opened-ports --format=json", but it's not really
+        # structured; it's just an array of strings which are the lines of the
+        # text output, like ["icmp","8081/udp"]. So it's probably just as
+        # likely to change as the text output, and doesn't seem any better.
+        output = typing.cast(str, self._run('opened-ports', return_output=True))
+        ports: Set[OpenedPort] = set()
+        for line in output.splitlines():
+            line = line.strip()
+            if not line:
+                continue
+            port = self._parse_opened_port(line)
+            if port is not None:
+                ports.add(port)
+        return ports
+
+    @classmethod
+    def _parse_opened_port(cls, port_str: str) -> Optional[OpenedPort]:
+        if port_str == 'icmp':
+            return OpenedPort('icmp', None)
+        port_range, slash, protocol = port_str.partition('/')
+        if not slash or protocol not in ['tcp', 'udp']:
+            logger.warning('Unexpected opened-ports protocol: %s', port_str)
+            return None
+        port, hyphen, _ = port_range.partition('-')
+        if hyphen:
+            logger.warning('Ignoring opened-ports port range: %s', port_str)
+        protocol_lit = typing.cast(typing.Literal['tcp', 'udp', 'icmp'], protocol)
+        return OpenedPort(protocol_lit, int(port))
```

Comment:
So this does clearly allow "icmp/10" but it does seem like we just filtered that out on line 3087.
Though I guess this is a type cast not a literal check?

Reply:
Fair point - fixed. (But yes, this is a type cast not a runtime check.)